frontend/src/metabase/common/components/Table/Table.tsx

```diff
@@ -1,69 +1,112 @@
-import React from "react";
+import type React from "react";
+import {
+  Fragment,
+  type PropsWithChildren,
+  useCallback,
+  useMemo,
+  useState,
+} from "react";
 
 import { color } from "metabase/lib/colors";
+import { useSelector } from "metabase/lib/redux";
+import { getLocale } from "metabase/setup/selectors";
 import { Box, Flex, Icon } from "metabase/ui";
 
-type BaseRow = Record<string, unknown> & { id: number | string };
+export type BaseRow = Record<string, any> & { id: number | string };
 
-type ColumnItem = {
+export type ColumnItem = {
   name: string;
   key: string;
+  sortable?: boolean;
 };
 
 export type TableProps<Row extends BaseRow> = {
   columns: ColumnItem[];
   rows: Row[];
   rowRenderer: (row: Row) => React.ReactNode;
-  tableProps?: React.HTMLAttributes<HTMLTableElement>;
-};
+  formatValueForSorting?: (row: Row, columnName: string) => any;
+  defaultSortColumn?: string;
+  defaultSortDirection?: "asc" | "desc";
+  ifEmpty?: React.ReactNode;
+} & React.HTMLAttributes<HTMLTableElement>;
 
 /**
  * A basic reusable table component that supports client-side sorting by a column
  *
- * @param columns     - an array of objects with name and key properties
- * @param rows        - an array of objects with keys that match the column keys
- * @param rowRenderer - a function that takes a row object and returns a <tr> element
- * @param tableProps  - additional props to pass to the <table> element
+ * @param {object} props
+ * @property columns - An array of objects with name and key properties
+ * @property rows - An array of row objects, which at minimum need an id
+ * @property rowRenderer - A function that takes a row object and returns a <tr> element
+ * @property formatValueForSorting
+ * @property defaultSortColumn
+ * @property defaultSortDirection
+ * @property ifEmpty - A component to render if the table is empty
+ * @note All other props are passed to the <table> element
  */
 export function Table<Row extends BaseRow>({
   columns,
   rows,
   rowRenderer,
+  formatValueForSorting = (row: Row, columnName: string) => row[columnName],
+  defaultSortColumn,
+  defaultSortDirection = "asc",
+  children,
+  ifEmpty = null,
   ...tableProps
-}: TableProps<Row>) {
-  const [sortColumn, setSortColumn] = React.useState<string | null>(null);
-  const [sortDirection, setSortDirection] = React.useState<"asc" | "desc">(
-    "asc",
+}: PropsWithChildren<TableProps<Row>>) {
+  const localeCode = useSelector(getLocale)?.code;
+
+  const [sortColumn, setSortColumn] = useState<string | undefined>(
+    defaultSortColumn,
+  );
+  const [sortDirection, setSortDirection] = useState<"asc" | "desc">(
+    defaultSortDirection,
+  );
+
+  const compareStrings = useCallback(
```

Comment:
It feels to me like this should be happening outside of the component. Again, probably doesn't matter for this iteration, but this table component is now tied to the redux store, which seems a little odd to me

Reply:
Good call. We could have a `locale` prop so `useLocale()` is called in the parent component and the result is passed into `Table`

Reply:
counterpoint: it's nice for `Table` to just handle that all by itself, rather than have to repeat the locale-retrieving logic in every component that calls `Table`